src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+                        valueToEncode = ch;
                     }
                     else
+#endif // ENTITY_ENCODE_HIGH_ASCII_CHARS
+                        if (char.IsSurrogate(ch))
                     {
-                        int valueToEncode = -1; // set to >= 0 if needs to be encoded
-
-#if ENTITY_ENCODE_HIGH_ASCII_CHARS
-                        if (ch >= 160 && ch < 256)
+                        int scalarValue = GetNextUnicodeScalarValueFromUtf16Surrogate(input, ref i);
+                        if (scalarValue >= UNICODE_PLANE01_START)
                         {
-                            // The seemingly arbitrary 160 comes from RFC
-                            valueToEncode = ch;
+                            valueToEncode = scalarValue;
                         }
                         else
-#endif // ENTITY_ENCODE_HIGH_ASCII_CHARS
-                        if (Char.IsSurrogate(ch))
                         {
-                            int scalarValue = GetNextUnicodeScalarValueFromUtf16Surrogate(ref pch, ref cch);
-                            if (scalarValue >= UNICODE_PLANE01_START)
-                            {
-                                valueToEncode = scalarValue;
-                            }
-                            else
-                            {
-                                // Don't encode BMP characters (like U+FFFD) since they wouldn't have
-                                // been encoded if explicitly present in the string anyway.
-                                ch = (char)scalarValue;
-                            }
+                            // Don't encode BMP characters (like U+FFFD) since they wouldn't have
+                            // been encoded if explicitly present in the string anyway.
+                            ch = (char)scalarValue;
                         }
+                    }
 
-                        if (valueToEncode >= 0)
-                        {
-                            // value needs to be encoded
-                            output.Append("&#");
-                            output.Append(valueToEncode.ToString(CultureInfo.InvariantCulture));
-                            output.Append(';');
-                        }
-                        else
-                        {
-                            // write out the character directly
-                            output.Append(ch);
-                        }
+                    if (valueToEncode >= 0)
+                    {
+                        // value needs to be encoded
+                        output.Append("&#");
+
+                        // Use the buffer directly and reserve a conservative estimate of 10 chars.
+                        Span<char> encodingBuffer = output.AppendSpan(10);
+                        valueToEncode.TryFormat(encodingBuffer, out int charsWritten, provider: CultureInfo.InvariantCulture);
+                        output.Length -= (10 - charsWritten);
```

Comment:
I think it'd be helpful to put the 10 that's used multiple times into a const that describes how it was picked, e.g. `const int MaxInt32Digits = 10` or something like that.  I assume valueToEncode can't be negative at this point?  If it could, max digits would need to be 11.